tests/unit/lib/SGMatrix_unittest.cc

```diff
@@ -761,3 +761,61 @@ TEST(SGMatrixTest,iterator)
 	for (auto v: mat)
 		EXPECT_EQ(mat[index++], v);
 }
+
+TEST(SGMatrixTest,column_iterator)
+{
+	constexpr index_t size = 5;
+	SGMatrix<float64_t> mat(size, size);
+	linalg::range_fill(mat, 1.0);
+
+	auto begin_col = mat.begin_column();
+	auto end_col = mat.end_column();
+
+	EXPECT_EQ(mat.num_cols, std::distance(begin_col, end_col));
+	EXPECT_EQ(mat.get_column(0), *begin_col);
+	++begin_col;
+	EXPECT_EQ(mat.get_column(1), *begin_col);
+	--begin_col;
+	EXPECT_EQ(mat.get_column(0), *begin_col++);
+	EXPECT_TRUE(begin_col != end_col);
+	++begin_col;
+	EXPECT_EQ(mat.get_column(2), *begin_col--);
+	--begin_col;
+	EXPECT_EQ(mat.get_column(0), *begin_col);
+	begin_col += 2;
+	EXPECT_EQ(mat.get_column(2), *begin_col);
+	begin_col -= 2;
+	EXPECT_EQ(mat.get_column(0), *begin_col);
+	EXPECT_EQ(mat.get_column(1), begin_col[1]);
+
+	auto new_itr = begin_col + 2;
+	EXPECT_EQ(mat.get_column(2), *new_itr);
+
+	// range-based loop should work as well
+	auto index = 0;
+	for (const auto& i : mat.columns())
+		EXPECT_EQ(i, mat.get_column(index++));
+
+
+	const SGMatrix<float64_t> const_mat(mat);
+
+	index = 0;
+	for (const auto& i : const_mat.columns())
+		EXPECT_EQ(i, const_mat.get_column(index++));
+
+	// Modifying value of column vector
+	auto mat_copy = mat.clone();
+
+	index = 0;
+	for (const auto& vec : mat_copy.columns())
+		mat_copy.set_column(index++, linalg::add(vec, vec));
+
+	index = 0;
+	for (const auto& vec : mat.columns())
+		mat.set_column(index++, linalg::add(vec,vec));
+
+	auto idx1 = 0 ,idx2 =0;
+	while(index < std::distance(begin_col,end_col))
+		EXPECT_EQ(mat.get_column(idx1++),mat_copy.get_column(idx2++));
```

Comment:
I think you missed the point of this test... You are now doing twice exactly the same and checking if the output is the same (which is a pretty basic assumption and I don't think we need to test for).
The test I suggested was to check whether updating the values in a loop was actually working when using the iterator. Please read the comment with the code snippet. What I meant was:
1. Mutate the vector from the iterator of a matrix
2. Compare in a separate loop so that we can be sure that mutating the vector returned by the iterator is actually changing the matrix itself and not a local copy. Hence you should either have a matrix with reference results or compute a reference value in a different way (e.g. do the same operation as you did in 1 with a local copy and compare that to the test matrix)